File: src/main/java/com/milaboratory/mixcr/cli/CommandClonesDiff.java
package com.milaboratory.mixcr.cli;

import com.milaboratory.core.sequence.NucleotideSequence;
import com.milaboratory.mixcr.basictypes.Clone;
import com.milaboratory.mixcr.basictypes.CloneSet;
import com.milaboratory.mixcr.basictypes.CloneSetIO;
import com.milaboratory.mixcr.basictypes.IOUtil;
import io.repseq.core.GeneType;
import io.repseq.core.VDJCGeneId;
import picocli.CommandLine.Command;
import picocli.CommandLine.Option;
import picocli.CommandLine.Parameters;

import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.PrintStream;
import java.util.*;

@Command(name = "clonesDiff",
        separator = " ",
        sortOptions = true,
        description = "Calculates the difference between two .clns files.")
public class CommandClonesDiff extends ACommandWithOutputMiXCR {
    @Parameters(description = "input1.clns")
    public String in1;

    @Parameters(description = "input2.clns")
    public String in2;

    @Parameters(description = "[report]", arity = "0..1")
    public String report = null;

    @Option(names = {"-v"}, description = "Use V gene in clone comparison (include it as a clone key along " +
            "with a clone sequence).")
    public boolean v = false;

    @Option(names = {"-j"}, description = "Use J gene in clone comparison (include it as a clone key along " +
            "with a clone sequence).")
    public boolean j = false;

    @Option(names = {"-c"}, description = "Use C gene in clone comparison (include it as a clone key along " +
            "with a clone sequence).")
    public boolean c = false;

    public boolean useV() { return v; }

    public boolean useJ() { return j; }

    public boolean useC() { return c; }

    //@Parameter(names = {"-o1", "--only-in-first"}, description = "output for alignments contained only " +
    //        "in the first .vdjca file")
    //public String onlyFirst;
    //@Parameter(names = {"-o2", "--only-in-second"}, description = "output for alignments contained only " +
    //        "in the second .vdjca file")
    //public String onlySecond;
    //@Parameter(names = {"-d1", "--diff-from-first"}, description = "output for alignments from the first file " +
    //        "that are different from those alignments in the second file")
    //public String diff1;
    //@Parameter(names = {"-d2", "--diff-from-second"}, description = "output for alignments from the second file " +
    //        "that are different from those alignments in the first file")
    //public String diff2;
    //@Parameter(names = {"-g", "--gene-feature"}, description = "Gene feature to compare")
    //public String geneFeatureToMatch = "CDR3";
    //@Parameter(names = {"-l", "--top-hits-level"}, description = "Number of top hits to search for match")
    //public int hitsCompareLevel = 1;


    @Override
    protected List<String> getOutputFiles() {
        return report == null ? Collections.emptyList() : Collections.singletonList(report);
    }

    @Override
    public void run0() throws Exception {
        try (InputStream is1 = IOUtil.createIS(in1);
             InputStream is2 = IOUtil.createIS(in2);
             PrintStream report = this.report == null ? System.out : new PrintStream(new FileOutputStream(this.report))) {

            CloneSet cs1 = CloneSetIO.readClns(is1);
            CloneSet cs2 = CloneSetIO.readClns(is2);

            HashMap<CKey, CRec> recs = new HashMap<>();

            populate(recs, cs1, 0);
            populate(recs, cs2, 1);

            int newClones1 = 0, newClones2 = 0;
            long newClones1Reads = 0, newClones2Reads = 0;

            for (CRec cRec : recs.values()) {
                if (cRec.clones[0] == null) {
                    newClones2++;
                    newClones2Reads += cRec.clones[1].getCount();
                }

                if (cRec.clones[1] == null) {
                    newClones1++;
                    newClones1Reads += cRec.clones[0].getCount();
                }
            }

            report.println("Unique clones in cloneset 1: " + newClones1 + " (" + Util.PERCENT_FORMAT.format(100.0 * newClones1 / cs1.size()) + "%)");
            report.println("Reads in unique clones in cloneset 1: " + newClones1Reads + " (" + Util.PERCENT_FORMAT.format(100.0 * newClones1Reads / cs1.getTotalCount()) + "%)");

            report.println("Unique clones in cloneset 2: " + newClones2 + " (" + Util.PERCENT_FORMAT.format(100.0 * newClones2 / cs2.size()) + "%)");
            report.println("Reads in unique clones in cloneset 2: " + newClones2Reads + " (" + Util.PERCENT_FORMAT.format(100.0 * newClones2Reads / cs2.getTotalCount()) + "%)");
        }
    }

    private void populate(Map<CKey, CRec> recs, CloneSet cs, int i) {
        for (Clone clone : cs) {
            CKey key = getKey(clone);
            CRec cRec = recs.get(key);
            if (cRec == null)
                recs.put(key, cRec = new CRec());

            if (cRec.clones[i] != null) {
                String error = "";
                char letter = 'X';
                if (!Objects.equals(
                        getBestGene(cRec.clones[i], GeneType.Variable),
                        getBestGene(clone, GeneType.Variable)))
                    letter = 'v';
                if (!Objects.equals(
                        getBestGene(cRec.clones[i], GeneType.Joining),
                        getBestGene(clone, GeneType.Joining)))
                    letter = 'j';
                if (!Objects.equals(
                        getBestGene(cRec.clones[i], GeneType.Constant),
                        getBestGene(clone, GeneType.Constant)))
                    letter = 'c';

                if (letter != 'X')
                    error = "Error: clones with the same key present in one of the clonesets. Seems that clones were assembled " +
                            "using -OseparateBy" + Character.toUpperCase(letter) + "=true option, please add -" + letter + " option to this command.";

                throwValidationException(error);
            }

            cRec.clones[i] = clone;
        }
    }

    private VDJCGeneId getBestGene(Clone clone, GeneType geneType) {
        return clone.getBestHit(geneType) == null ? null : clone.getBestHit(geneType).getGene().getId();
    }

    private CKey getKey(Clone clone) {
        final NucleotideSequence[] clonalSequence = new NucleotideSequence[clone.numberOfTargets()];
        for (int i = 0; i < clonalSequence.length; i++)
            clonalSequence[i] = clone.getTarget(i).getSequence();

        final VDJCGeneId v = useV() ? getBestGene(clone, GeneType.Variable) : null;

        final VDJCGeneId j = useJ() ? getBestGene(clone, GeneType.Joining) : null;

        final VDJCGeneId c = useC() ? getBestGene(clone, GeneType.Constant) : null;

        return new CKey(clonalSequence, v, j, c);
    }

    private static final class CRec {
        final Clone[] clones = new Clone[2];
    }

    private static final class CKey {
        final NucleotideSequence[] clonalSequence;
        final VDJCGeneId v, j, c;

        public CKey(NucleotideSequence[] clonalSequence, VDJCGeneId v, VDJCGeneId j, VDJCGeneId c) {
            this.clonalSequence = clonalSequence;
            this.v = v;
            this.j = j;
            this.c = c;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof CKey)) return false;

            CKey cKey = (CKey) o;

            // Probably incorrect - comparing Object[] arrays with Arrays.equals
            if (!Arrays.equals(clonalSequence, cKey.clonalSequence)) return false;
            if (v != null ? !v.equals(cKey.v) : cKey.v != null) return false;
            if (j != null ? !j.equals(cKey.j) : cKey.j != null) return false;
            return c != null ? c.equals(cKey.c) : cKey.c == null;

        }

        @Override
        public int hashCode() {
            int result = Arrays.hashCode(clonalSequence);
            result = 31 * result + (v != null ? v.hashCode() : 0);
            result = 31 * result + (j != null ? j.hashCode() : 0);
            result = 31 * result + (c != null ? c.hashCode() : 0);
            return result;
        }
    }
}

File: src/main/kotlin/com/milaboratory/mixcr/cli/CommandClonesDiff.kt
/*
 * Copyright (c) 2014-2023, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.cli

import com.milaboratory.app.InputFileType
import com.milaboratory.app.ValidationException
import com.milaboratory.core.sequence.NucleotideSequence
import com.milaboratory.mixcr.basictypes.Clone
import com.milaboratory.mixcr.basictypes.CloneSet
import com.milaboratory.mixcr.basictypes.CloneSetIO
import com.milaboratory.util.ReportHelper
import io.repseq.core.GeneType
import io.repseq.core.GeneType.Constant
import io.repseq.core.GeneType.Joining
import io.repseq.core.GeneType.Variable
import io.repseq.core.VDJCGeneId
import picocli.CommandLine.Command
import picocli.CommandLine.Option
import picocli.CommandLine.Parameters
import java.io.PrintStream
import java.nio.file.Path

@Command(
    description = ["Calculates the difference between two .clns files."]
)
class CommandClonesDiff : MiXCRCommandWithOutputs() {
    @Parameters(paramLabel = "input1.(clns|clna)", index = "0")
    lateinit var in1: Path

    @Parameters(paramLabel = "input2.(clns|clna)", index = "1")
    lateinit var in2: Path

    @Parameters(
        description = ["Path where to write report. Will write to output if omitted."],
        paramLabel = "report.txt",
        index = "2",
        arity = "0..1"
    )
    var report: Path? = null

    @Option(
        names = ["-v", "--use-v"],
        description = ["Use V gene in clone comparison (include it as a clone key along with a clone sequence)."],
        order = OptionsOrder.main + 10_100
    )
    var useV = false

    @Option(
        names = ["-j", "--use-j"],
        description = ["Use J gene in clone comparison (include it as a clone key along with a clone sequence)."],
        order = OptionsOrder.main + 10_200
    )
    var useJ = false

    @Option(
        names = ["-c", "--use-cc"],
        description = ["Use C gene in clone comparison (include it as a clone key along with a clone sequence)."],
        order = OptionsOrder.main + 10_300
    )
    var useC = false

    //@Parameter(names = {"-o1", "--only-in-first"}, description = "output for alignments contained only " +
    //        "in the first .vdjca file")
    //public String onlyFirst;
    //@Parameter(names = {"-o2", "--only-in-second"}, description = "output for alignments contained only " +
    //        "in the second .vdjca file")
    //public String onlySecond;
    //@Parameter(names = {"-d1", "--diff-from-first"}, description = "output for alignments from the first file " +
    //        "that are different from those alignments in the second file")
    //public String diff1;
    //@Parameter(names = {"-d2", "--diff-from-second"}, description = "output for alignments from the second file " +
    //        "that are different from those alignments in the first file")
    //public String diff2;
    //@Parameter(names = {"-g", "--gene-feature"}, description = "Gene feature to compare")
    //public String geneFeatureToMatch = "CDR3";
    //@Parameter(names = {"-l", "--top-hits-level"}, description = "Number of top hits to search for match")
    //public int hitsCompareLevel = 1;
    override val inputFiles
        get() = listOf(in1, in2)

    override val outputFiles
        get() = listOfNotNull(report)

    override fun validate() {
        ValidationException.requireFileType(in1, InputFileType.CLNX)
        ValidationException.requireFileType(in2, InputFileType.CLNX)
        ValidationException.requireFileType(report, InputFileType.TXT)
    }

    override fun run1() {
        (report?.let { PrintStream(it.toFile()) } ?: System.out).use { report ->
            val cs1 = CloneSetIO.read(in1)
            val cs2 = CloneSetIO.read(in2)
            val recs = mutableMapOf<CKey, CRec>()
            populate(recs, cs1, 0)
            populate(recs, cs2, 1)
            var newClones1 = 0
            var newClones2 = 0
            var newClones1Reads: Long = 0
            var newClones2Reads: Long = 0
            for (cRec in recs.values) {
                if (cRec.clones[0] == null) {
                    newClones2++
                    newClones2Reads += cRec.clones[1]!!.count.toLong()
                }
                if (cRec.clones[1] == null) {
                    newClones1++
                    newClones1Reads += cRec.clones[0]!!.count.toLong()
                }
            }
            report.println(
                "Unique clones in cloneset 1: $newClones1 " +
                        "(${ReportHelper.PERCENT_FORMAT.format(100.0 * newClones1 / cs1.size())}%)"
            )
            report.println(
                "Reads in unique clones in cloneset 1: $newClones1Reads " +
                        "(${ReportHelper.PERCENT_FORMAT.format(100.0 * newClones1Reads / cs1.totalCount)}%)"
            )
            report.println(
                "Unique clones in cloneset 2: $newClones2 " +
                        "(${ReportHelper.PERCENT_FORMAT.format(100.0 * newClones2 / cs2.size())}%)"
            )
            report.println(
                "Reads in unique clones in cloneset 2: $newClones2Reads " +
                        "(${ReportHelper.PERCENT_FORMAT.format(100.0 * newClones2Reads / cs2.totalCount)}%)"
            )
        }
    }

    private fun populate(recs: MutableMap<CKey, CRec>, cs: CloneSet, i: Int) {
        for (clone in cs) {
            val key = getKey(clone)
            val cRec = recs.computeIfAbsent(key) { CRec() }
            cRec.clones[i]?.let { unexpectedClone ->
                val letter = when {
                    getBestGene(unexpectedClone, Constant) != getBestGene(clone, Constant) -> 'c'
                    getBestGene(unexpectedClone, Joining) != getBestGene(clone, Joining) -> 'j'
                    getBestGene(unexpectedClone, Variable) != getBestGene(clone, Variable) -> 'v'
                    else -> 'X'
                }
                val error: String = when {
                    letter != 'X' -> "Error: clones with the same key present in one of the clonesets. Seems that clones were assembled " +
                            "using -OseparateBy${letter.uppercaseChar()}=true option, please add -$letter option to this command."
                    else -> ""
                }
                throw ValidationException(error)
            }
            cRec.clones[i] = clone
        }
    }

    private fun getBestGene(clone: Clone, geneType: GeneType): VDJCGeneId? {
        return clone.getBestHit(geneType)?.gene?.id
    }

    private fun getKey(clone: Clone): CKey {
        val clonalSequence = Array<NucleotideSequence>(clone.numberOfTargets()) { i ->
            clone.getTarget(i).sequence
        }
        val v = if (useV) getBestGene(clone, Variable) else null
        val j = if (useJ) getBestGene(clone, Joining) else null
        val c = if (useC) getBestGene(clone, Constant) else null
        return CKey(clonalSequence, v, j, c)
    }

    private class CRec {
        val clones = arrayOfNulls<Clone>(2)
    }

    private class CKey(
        val clonalSequence: Array<NucleotideSequence>,
        val v: VDJCGeneId?,
        val j: VDJCGeneId?,
        val c: VDJCGeneId?
    ) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false

            other as CKey

            if (!clonalSequence.contentEquals(other.clonalSequence)) return false
            if (v != other.v) return false
            if (j != other.j) return false
            if (c != other.c) return false

            return true
        }

        override fun hashCode(): Int {
            var result = clonalSequence.contentHashCode()
            result = 31 * result + (v?.hashCode() ?: 0)
            result = 31 * result + (j?.hashCode() ?: 0)
            result = 31 * result + (c?.hashCode() ?: 0)
            return result
        }
    }
}
